### PR TITLE
feat(browser): add browser.disconnected event

### LINKED ISSUE #952: Add `browser.closed` and `page.closed` events
- Fire `browser.closed` event when the underlying browser process gets closed / crashes
- Fire `page.closed` event when the page gets closed / crashes

This will help to build reliable scripts that restart browser or reopen pages when they crash.

### PR DESCRIPTION
**Pull request by sukrosono**
This patch add browser.disconnected event :
- [x] when browser.close() called.
- [x] when browser.disconnect() called.
fix #952

### CODE REVIEW

**Comment by sukrosono** [comment]
this just take one point on [existing issues](https://github.com/GoogleChrome/puppeteer/issues/952).
When the other one still open i will take care of it when i got free time.

#### Out of context 
My attempt was add beforeEach block under Page.close suite, but the linting yell `Unexpected token browser` with plain function or arrow syntax. So i giving up on that.

File: docs/api.md [hunk 1]
```diff
@@ -254,6 +255,9 @@ puppeteer.launch().then(async browser => {
 });
 ```
 
+#### event: 'closed'
+Emitted when page closed or crashed.
```
Comment (aslushnikov) [inline docs/api.md]: nit: Emitted when page gets closed or page renderer crashes.

File: lib/Page.js [hunk 2]
```diff
@@ -663,8 +664,13 @@ class Page extends EventEmitter {
     return this.mainFrame().title();
   }
 
+  _closed() {
```
Comment (aslushnikov) [inline lib/Page.js]: let's inline this method

File: test/test.js [hunk 3]
```diff
@@ -184,6 +184,26 @@ describe('Page', function() {
...
+      await newPage.close();
+      expect(fired).toBe(true);
+    }));
+  });
+
+  describe('Page.Events.closed', function(){
+    it('should emmited when page got crashed', SX(async function() {
+      let emitted = false;
+      page.on('closed', () => emitted = true);
+      page.goto('chrome://crash');
```
Comment (aslushnikov) [inline test/test.js]: you should be able to await `page.goto` here, instead of doing the `waitForEvents` on the following line

Comment (sukrosono) [reply test/test.js]: it fail on my machine if i do `await page.goto(url);` , with jasmine timeout.

Comment (sukrosono) [reply test/test.js]: [here](https://travis-ci.org/GoogleChrome/puppeteer/jobs/284151708#L525) the travis test failure. That's why i wait on error, maybe you can assist me?

**Comment by sukrosono** [comment]
@aslushnikov dude, i need help on browser.closed, when it crashes. I have no idea how to make the browser crashes and how to make spec about it. I already trying a couple of hour but no result.

**Comment by ebidel** [comment]
@brutalcrozt the page `chrome://crash/` will crash the browser. you can use that for testing!

**Comment by sukrosono** [comment]
@ebidel Thanks Eric, i notice that now. But surely i still don't know how to make `browser._onCrash`. Maybe that because my lack of understanding the relation browser and page in the codebase.

**Comment by sukrosono** [comment]
Since now we get `Browser.pages` and `Browser.targets` , 

`Browser.closed` now should only fired when get closed via `Browser.close`.
ping @aslushnikov  for review, fix #952

File: lib/Browser.js [hunk 4]
```diff
@@ -122,16 +122,25 @@ class Browser extends EventEmitter {
...
+    this._closed();
   }
 
   disconnect() {
     this._connection.dispose();
   }
 }
 
+Browser.Events = {
+  Closed: 'closed'
```
Comment (aslushnikov) [inline lib/Browser.js]: This should be called `'disconnected'` and should be emitted whenever the connection is closed

File: lib/Page.js [hunk 5]
```diff
@@ -98,6 +98,7 @@ class Page extends EventEmitter {
   }
 
   _onTargetCrashed() {
+    this._closed();
```
Comment (aslushnikov) [inline lib/Page.js]: there should be no need to emit the error event now; it should be just 'crashed'

cc @JoelEinbinder

File: lib/Page.js [hunk 6]
```diff
@@ -728,8 +729,11 @@ class Page extends EventEmitter {
     return this.mainFrame().title();
   }
 
+  _closed() { this.emit(Page.Events.Closed, 'page closed'); }
```
Comment (aslushnikov) [inline lib/Page.js]: the 'closed' event is not needed any more since we have target events; they should cover everything

File: lib/Browser.js (hunk 4 above)
Comment (sukrosono) [reply lib/Browser.js]: Done, thanks for put a trust in me.

**Comment by sukrosono** [comment]
@aslushnikov  IDK why this PR fail on node 7 when running yarn run lint.

File: lib/Browser.js [hunk 7]
```diff
@@ -122,21 +122,27 @@ class Browser extends EventEmitter {
...
+  _disconnected() { this.emit(Browser.Events.Disconnected, 'connection disconnected');}
+
   async close() {
     await this._closeCallback.call(null);
     this.disconnect();
   }
 
   disconnect() {
     this._connection.dispose();
+    this._disconnected();
```
Comment (aslushnikov) [inline lib/Browser.js]: the event should also be fired whenever the connection itself gets disconnected (e.g. whenever the browser gets closed)

File: lib/Browser.js [hunk 8]
```diff
@@ -122,21 +122,27 @@ class Browser extends EventEmitter {
     return version.product;
   }
 
+  _disconnected() { this.emit(Browser.Events.Disconnected, 'connection disconnected');}
```
Comment (aslushnikov) [inline lib/Browser.js]: nit: formatting here is improper. The "inlining" is to remove the function `_disconnected` and just emit the event in `disconnected` method right away.

Comment (sukrosono) [reply lib/Browser.js]: done

File: lib/Browser.js (hunk 7 above)
Comment (sukrosono) [reply lib/Browser.js]: @aslushnikov Sorry, i don't get it. Could you re-explain it to me? :confused:

**Comment by sukrosono** [comment]
ping @aslushnikov  for review

File: lib/Browser.js [hunk 9]
```diff
@@ -122,21 +122,26 @@ class Browser extends EventEmitter {
...
 
+
   async close() {
     await this._closeCallback.call(null);
     this.disconnect();
   }
 
   disconnect() {
     this._connection.dispose();
+    this.emit(Browser.Events.Disconnected, 'connection disconnected');
```
Comment (aslushnikov) [inline lib/Browser.js]: this event should be emitted when the browser gets closed; e.g. when it crashes.

Comment (sukrosono) [reply lib/Browser.js]: Done, let me know if it still wrong. So it's introduce breaking changes?

Comment (aslushnikov) [reply lib/Browser.js]: Wait, I think there's some miscommunication here.

I actually meant the browser crash, not the page crash.

The only signal we get when the browser crashes is the disconnect of web socket (see lib/Connection.js). We should capture this event and dispatch the `disconnected` for it

Comment (sukrosono) [reply lib/Browser.js]: Okay, we take a look on it. I also need your advice to reflect the documentation changes for this one. FYI build fail because travis are down.

Comment (sukrosono) [reply lib/Browser.js]: I hope that was the right one, i skip the test for this case.

File: test/test.js [hunk 10]
```diff
@@ -235,6 +239,33 @@ describe('Page', function() {
...
+        browserWSEndpoint: originalBrowser.wsEndpoint()
+      });
+      let isDisconnected = false;
+      newBrowser.on('disconnected', dc => {
+        isDisconnected = true;
+      });
+      newBrowser.disconnect();
+      expect(isDisconnected).toBe(true);
+    }));
+    xit('should be emitted when underlying websocket gets closed', SX(async function() {
```
Comment (aslushnikov) [inline test/test.js]: you should be able to:
1. launch a browser
2. connect to it with `pptr.connect`
2. close the browser, make sure the connected one issues the `disconnected` event

File: lib/Browser.js [hunk 11]
```diff
@@ -122,21 +125,26 @@ class Browser extends EventEmitter {
     return version.product;
   }
 
+
```
Comment (aslushnikov) [inline lib/Browser.js]: nit: stray line

File: lib/Browser.js [hunk 12]
```diff
@@ -33,6 +33,9 @@ class Browser extends EventEmitter {
     this._closeCallback = closeCallback || new Function();
     /** @type {Map<string, Target>} */
     this._targets = new Map();
+    this._connection.on('ws.closed', closed => {
+      this.disconnect();
```
Comment (aslushnikov) [inline lib/Browser.js]: no need to run `this.disconnect()` here - you only need to emit the event.

File: test/test.js (hunk 10 above)
Comment (sukrosono) [reply test/test.js]: Let me know if it still not fit, is that a parrot in your profile picture?

File: docs/api.md [hunk 13]
```diff
@@ -288,6 +289,8 @@ puppeteer.launch().then(async browser => {
   await browser2.close();
 });
 ```
+#### event: 'disconnected'
+emitted whenever the connection is closed `Browser.close() Browser.disconnect()`.
```
Comment (sukrosono) [inline docs/api.md]: Not sure this one, let me know if any better suggestion.

File: lib/Browser.js (hunk 12 above)
Comment (sukrosono) [reply lib/Browser.js]: done

File: lib/Browser.js (hunk 11 above)
Comment (sukrosono) [reply lib/Browser.js]: done

File: docs/api.md (hunk 13 above)
Comment (aslushnikov) [inline docs/api.md]: Let's do the following wording:

> Emitted when puppeteer gets disconnected from the browser instance. This might happen because one of the following:
> - browser closed or crashed
> - `browser.disconnect` method was called

File: lib/Browser.js [hunk 14]
```diff
@@ -33,6 +33,9 @@ class Browser extends EventEmitter {
     this._closeCallback = closeCallback || new Function();
     /** @type {Map<string, Target>} */
     this._targets = new Map();
+    this._connection.on('ws.closed', closed => {
```
Comment (aslushnikov) [inline lib/Browser.js]: we shouldn't whitelist certain events and give them special treatment.

Instead, let's introduce a "closedCallback" for the connection and initialize it here:

```js
this._connection.setClosedCallback(() => {
  this.emit(Browser.Events.Disconnected);
});
```

File: lib/Browser.js [hunk 15]
```diff
@@ -33,6 +33,9 @@ class Browser extends EventEmitter {
...
+      this.emit(Browser.Events.Disconnected, 'connection disconnected');
```
Comment (aslushnikov) [inline lib/Browser.js]: let's drop the argument for the event, it's unclear if it's useful.

File: lib/Browser.js [hunk 16]
```diff
@@ -129,14 +132,16 @@ class Browser extends EventEmitter {
 
   disconnect() {
     this._connection.dispose();
+    this.emit(Browser.Events.Disconnected, 'connection disconnected');
```
Comment (aslushnikov) [inline lib/Browser.js]: no need to emit here: this will result in disconnected event dispatched twice (one from here, another from the connection been closed)

File: lib/Connection.js [hunk 17]
```diff
@@ -107,7 +107,8 @@ class Connection extends EventEmitter {
     }
   }
 
-  _onClose() {
+  _onClose(code, reason) {
+    if (code) this.emit('ws.closed', [code, reason]);
```
Comment (aslushnikov) [inline lib/Connection.js]: 1. let's emit as the very last line in the method. This way, if there's an exception in user-land code, we won't be left in a non-consistent state.
2. We should emit unconditionally
3. Let's not emit any details for now

File: test/test.js [hunk 18]
```diff
@@ -235,6 +235,46 @@ describe('Page', function() {
...
+      const originalBrowser = await puppeteer.launch(defaultBrowserOptions);
+      const newBrowser = await puppeteer.connect({
+        browserWSEndpoint: originalBrowser.wsEndpoint()
+      });
+      let isDisconnected = false;
+      newBrowser.on('disconnected', dc => {
+        isDisconnected = true;
+      });
+      newBrowser.disconnect();
+      expect(isDisconnected).toBe(true);
```
Comment (aslushnikov) [inline test/test.js]: please, close originalBrowser here so that it's not left dangling

File: test/test.js [hunk 19]
```diff
@@ -235,6 +235,46 @@ describe('Page', function() {
...
+      let resolver = null;
+      const promise = new Promise((res, rej) => {
+        resolver = res;
+      });
+      let isDisconnected = false;
+      browser.on('disconnected', dc => {
+        isDisconnected = true;
+        resolver();
+      });
+      browser._connection._ws.close(3006, 'intentionally kill it');
```
Comment (aslushnikov) [inline test/test.js]: we should not reach to private variables in tests.
Instead, this test could be written like the previous one, but instead of disconnecting
the new browser, closing original one:

```js
it('should be emitted when underlying websocket gets closed', SX(async function() {
  const browser = await puppeteer.launch(defaultBrowserOptions);
  const newBrowser = await puppeteer.connect({
    browserWSEndpoint: originalBrowser.wsEndpoint()
  });
  let isDisconnected = false;
  newBrowser.once('disconnected', () => isDisconnected = true);
  await originalBrowser.close();
  expect(isDisconnected).toBe(true);
}));

```

**Review by aslushnikov — COMMENTED** [review]
We're getting close!

File: lib/Connection.js [hunk 20]
```diff
@@ -1,217 +1,222 @@
...
-  _onClosed() {
-    for (const callback of this._callbacks.values())
-      callback.reject(new Error(`Protocol error (${callback.method}): Target closed.`));
-    this._callbacks.clear();
-    this._connection = null;
-  }
-}
-
-module.exports = {Connection, Session};
+/**
```
Comment (aslushnikov) [inline lib/Connection.js]: Are these different line-endings? Can you please revert the change so that the actual changes are clear?

File: test/test.js (hunk 19 above)
Comment (sukrosono) [reply test/test.js]: Sorry i didn't know that, i just try to get closer to 
> should be emitted when underlying websocket gets closed

Most of my time yesterday was spend here, i wish know it earlier. Next time i should be more clever on testing.

File: lib/Connection.js (hunk 17 above)
Comment (sukrosono) [reply lib/Connection.js]: done, hopefully i am not misunderstand again.

File: test/test.js (hunk 19 above)
Comment (aslushnikov) [reply test/test.js]: @brutalcrozt feel free to ask questions next time if you're getting stuck.

File: lib/Connection.js (hunk 20 above)
Comment (sukrosono) [reply lib/Connection.js]: Perhaps, certainly i will do.

File: test/test.js [hunk 21]
```diff
@@ -235,6 +235,37 @@ describe('Page', function() {
     }));
   });
 
+  describe('Browser.Events.disconnected', function() {
+    it('should emitted when browser gets closed', SX(async function() {
```
Comment (sukrosono) [inline test/test.js]: this one fired `disconnected` twice, i notice that. Same effect when i use the `browser`, Assuming it added to two instance the browser and originalBrowser. For me it's important you to know. If i am wrong please leave some suggestion.

**Comment by sukrosono** [comment]
@aslushnikov This one is ready captain. :smile:

File: lib/Connection.js [hunk 22]
```diff
@@ -76,6 +76,11 @@ class Connection extends EventEmitter {
   }
 
   /**
+   * @param {function()} callback
+   */
+  setClosedCallback(callback) { this._closeCallback = callback; }
```
Comment (aslushnikov) [inline lib/Connection.js]: can you please format this?
```js
setClosedCallback(callback) {
  this._closeCallback = callback;
}
```

File: lib/Connection.js [hunk 23]
```diff
@@ -108,6 +113,7 @@ class Connection extends EventEmitter {
   }
 
   _onClose() {
+    this._closeCallback();
```
Comment (aslushnikov) [inline lib/Connection.js]: 1. this should be the last line in the method
2. for the sake of cleanliness, let's check that delegate is set:
```js
if (this._closeCallback)
  this._closeCallback();
```

File: test/test.js (hunk 21 above)
Comment (aslushnikov) [reply test/test.js]: Good catch!

That's because:
1. Connection should reset it's `this._ws` to `null` once the COnnection._onClose method is called.
2. Connection._onClose should remove added event listeners (use `helper.addEventListener` / `helper.removeEventListeners` for this)
2. The Connection.dispose() method should fast-return if this._ws === null.

File: test/test.js [hunk 24]
```diff
@@ -235,6 +235,37 @@ describe('Page', function() {
...
+      const originalBrowser = await puppeteer.launch(defaultBrowserOptions);
+      let isDisconnected = false;
+      originalBrowser.on('disconnected', () => { isDisconnected = true; });
+      await originalBrowser.close();
+      expect(isDisconnected).toBe(true);
+    }));
+    it('should emitted when browser.disconnect() called', SX(async function() {
```
Comment (aslushnikov) [inline test/test.js]: Let's unify these three tests into one. This will save us on some time while running tests.

```js
const originalBrowser = await puppeteer.launch(defaultBrowserOptions);
const browserWSEndpoint = originalBrowser.wsEndpoint();
const remoteBrowser1 = await puppeteer.connect({browserWSEndpoint});
const remoteBrowser2 = await puppeteer.connect({browserWSEndpoint});

let disconnectedOriginal = 0;
let disconnectedRemote1 = 0;
let disconnectedRemote2 = 0;
originalBrowser.on('disconnected', () => ++disconnectedOriginal);
remoteBrowser1.on('disconnected', () => ++disconnectedRemote1);
remoteBrowser2.on('disconnected', () => ++disconnectedRemote2);

await remoteBrowser2.disconnect();
expect(disconnectedOriginal).toBe(0);
expect(disconnectedRemote1).toBe(0);
expect(disconnectedRemote2).toBe(1);

await originalBrowser.close();
expect(disconnectedOriginal).toBe(1);
expect(disconnectedRemote1).toBe(1);
expect(disconnectedRemote2).toBe(1);

File: lib/Connection.js (hunk 22 above)
Comment (sukrosono) [reply lib/Connection.js]: Done, so no inline on class method? Inline only on normal code?

File: test/test.js (hunk 21 above)
Comment (sukrosono) [reply test/test.js]: I am not fully understand it, but i do the trick. Hopefully good to go :smile:

File: lib/Connection.js (hunk 22 above)
Comment (aslushnikov) [reply lib/Connection.js]: "Inlining" is not about code formatting.
To "inline a function" means to get the function body and paste it instead of function invocations.

More on wiki: [inline expanstion](https://en.wikipedia.org/wiki/Inline_expansion)

**Review by aslushnikov — APPROVED** [review]

**Comment by aslushnikov** [comment]
Thanks!
